File: go/2015/day04a/day04a.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
)

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <key>\n", progname)
	os.Exit(1)
}

func process(key string) uint32 {
	for i := 1 ; ; i++ {
		tryKey := fmt.Sprintf("%s%d", key, i)
		hash := md5.Sum([]byte(tryKey))
		hashStr := hex.EncodeToString(hash[:])
		if hashStr[:5] == "00000" {
			return uint32(i)
		}
	}
	return 0
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: c/2015/day04a/day04a.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "md5.h"

void usage(char *progname) {
    fprintf(stderr, "usage: %s <input file>\n", progname);
    exit(1);
}

int process(char *key) {
    char try_key[32];
    char digest[33];
    int n = 1;

    for ( ; ; ) {
        sprintf(try_key, "%s%d", key, n);
	md5_digest(try_key, digest);
        if ( strncmp(digest, "00000", 5) == 0 ) {
            return n;
        }
	n++;
    }
  
}

int main(int argc, char *argv[]) {
    char *key;
    int result;

    if ( argc < 2 ) {
        usage(argv[0]);
    }

    key = argv[1];
    result = process(key);
    printf("result = %d\n", result);
    return 0;
}
